Implement basic metadata caching

AndrejMitrovic opened the pull request:
Might require some test-case as well. I'll consult with the test master @AndrejMitrovic if he wants me to write any.

AndrejMitrovic commented on source/agora/common/Metadata.d:
```diff
@@ -0,0 +1,109 @@
+///
+unittest
+{
+    auto meta = new Metadata(tempDir());
```
this might fail in Travis, I should probably just use a MemoryDevice buffer like we used to use in ocean.

AndrejMitrovic commented:
> std.file.FileException@std/file.d(1057): Failed to remove file /tmp/metadata.dat: No such file or directory

~I'll fix it up to use a memory device tomorrow.~

Edit: I went with the Appender approach because the Phobos File is a struct so it's not as easy to provide a custom "File" that writes to a buffer.

AndrejMitrovic commented:
It's green now.

Geod24 commented on doc/config.example.yaml:
```diff
@@ -37,6 +37,8 @@ node:
   # Public address:  GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ
   seed:    SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4
+  # Path to the data directory (if the path doesn't exist it will be created)
+  data_dir: /Users/you/agora/data
```
I'd put something more usable, like ".cache" so there are less items to modify.

Geod24 commented on source/agora/common/Metadata.d:
```diff
@@ -0,0 +1,132 @@
+
+    /// Load any existing metadata from disk (if it exists),
+    /// and set 'this.loaded' to true if at least one validator was loaded
+    public void load ( )
```
This class shouldn't do IO. Make it accept a string and take the IO out of it.

AndrejMitrovic replied:
Alright. I didn't think it was that important since it just stores IPs.

Geod24 commented on source/agora/common/Metadata.d:
```diff
@@ -0,0 +1,132 @@
+{
+    auto meta = new Metadata(tempDir());
+    const Validator = "http://127.0.0.1:1234";
+    const Listener = "http://127.0.0.1:4567";
```
Yeah that's not scalable. Since we serialize to `json`, you should be able to do:
```d
unittest
{
    auto meta = new Metadata("{ ... }");
}
```

The API here does not matter to much though. You might want to make your ctor non-validating and provide a `static Metadata fromFile(string path)` function as well as a `static Metadata fromString(string)`.

AndrejMitrovic replied:
Alrighty, I'll put these as helper functions outside the Metadata class.

Geod24 commented on source/agora/node/Network.d:
```diff
@@ -25,6 +26,7 @@ import libsodium.randombytes;
 
 import vibe.core.core;
 import vibe.core.log;
+import vibe.data.json;
```
why?

AndrejMitrovic replied:
Leftover code, will remove.

Geod24 reviewed (changes requested):
Need to take the IO out of the class, or at least make it more modular so that we don't have to provide a special `version` for testing purposes.

AndrejMitrovic commented:
> Need to take the IO out of the class, or at least make it more modular so that we don't have to provide a special version for testing purposes.

The thing is that just moves the IO responsibility somewhere else. So, who gets to do IO if it's not encapsulated in this class? Do I need to store a File handle in the Network class side-by-side with the Metadata class? Because that's dirty to me.

AndrejMitrovic commented:
How about something like this:

```D
class Metadata
{
     T ips;  // the data to dump
     Writer writer;  // in-memory or file-based writer

     this (Writer writer) { this.writer = writer; }
}
```

And then in the Network class we instantiate it either with a file-based writer or an in-memory writer based on whether we compile in -unittest mode?

Can you think of a better or cleaner way to do this? With Ocean/Tango it was a bit simpler since File is a subclass of an IODevice, and you can just drop in a MemoryDevice to write to memory rather than to disk.

AndrejMitrovic commented:
Ok now that you've made your PRs I know what to do here.

AndrejMitrovic commented on source/agora/node/Network.d:
```diff
@@ -257,4 +281,22 @@ class Network
+
+        Reads the metadata from the provided disk path.
+
+        Subclasses could override this method and return
```
So I made something similar to `getMetadata`. The unittest code will override this function and return a `new MemMetadata` rather than `new DiskMetadata`.

Now I just need your other PR to be finished so I can add some tests on top of it. @Geod24

AndrejMitrovic committed:
Implement use & serialization of metadata

It uses simple Json serializing without any
checksums or performance optimizations.

These routines will be replaced by SQLite or
another serializing library later.

AndrejMitrovic committed:
Provide a default unittest ctor in NetworkManager

This avoids having to pass "fake" arguments
to its constructor for the instance of NetworkManager
that the unittests use to keep track of all nodes.

AndrejMitrovic commented:
Updated PR. Feel free to destroy.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -258,9 +261,7 @@ public NetworkT makeTestNetwork (NetworkT : TestNetworkManager)
 
-
-        auto net = new NetworkT(NodeConfig.init, node_configs.map!(
-            c => c.key_pair.address.toString()).array);
```
I think that's where our approaches differ. I think that those arguments should be used / useful, and I thought they currently were, at least for the second one ?

Geod24 replied:
Well it makes sense in the context of your previous PR which only uses `apis`. But if we go this way, we do not need, nor do we want, to have a `NetworkManager`, because we're essentially conflating 2 use cases into one.

AndrejMitrovic replied:
Yeah that's why I suggested splitting up these into two different classes. But if keep them the same class, the thing is these arguments that we pass here are just not used at all in the test-suite. Which approach do we take?

Geod24 commented on source/agora/node/Network.d:
```diff
@@ -46,7 +47,7 @@ import std.random;
 {
     /// Config instance
-    private const NodeConfig node_config;
+    private const NodeConfig node_config = NodeConfig.init;
```
Wat

Geod24 replied:
I'm pretty sure this can be removed ?

Geod24 commented on source/agora/common/Metadata.d:
```diff
@@ -0,0 +1,137 @@
+    public Set!Address peers;
+
+    /// Deserialize the Metadata from a Json string
+    public static void fromJsonString (Metadata metadata, string json)
```
I'm not soooooo found of that design. Mixing `class` and JSON object usually doesn't end well.

The way I tend to look at it is that classes have an "identity" as two bitwise-equal objects are different, but value types have no identity as they are defined by the sum of their content.
For JSON object you want value type, not something with an identity.

AndrejMitrovic replied:
Well I can't use structs if I want to be able to provide Metadata instance that doesn't use IO and is used during unittests. What do you propose instead?

Geod24 commented on source/agora/test/Metadata.d:
```diff
@@ -0,0 +1,69 @@
+
+*******************************************************************************/
+
+module agora.test.Metadata;
```
Does it need its own test ? It should be more or less ignored in unittest for the time being.

AndrejMitrovic replied:
I'm confused. Are we doing TDD or not?

Geod24 replied:
We are but the point of abstracting away this is that it does nothing in unittests. So testing that it does nothing is a bit odd. In fact this test looks very much like the `Network` test.

AndrejMitrovic replied:
How do you make it do nothing in unittests then? You do one of these:

- Do a `version(unittest) { } else { this.metadata.loadFromDisk()` }`
- Subclass Metadata and then provide a File-based one by default, and an in-memory one in the unittests

Is there a third option? I really don't understand.

Geod24 reviewed (changes requested):
Needs a rebase and I think we can simplify it

AndrejMitrovic commented:
Rebased.

AndrejMitrovic commented:
@Geod24 ready?

Geod24 commented on source/agora/common/Metadata.d:
```diff
@@ -0,0 +1,137 @@
+}
+
+/// Metadata stored in memory (for unittests)
+public class MemMetadata : Metadata
```
I think this should live in the `Base` module.
One thing I really try to avoid is mixing test code with "regular code".
There's two reason for that:
- Better separation of concern. The "regular" code should not know about the test code, but the other way around is fine.
- The regular code should not have any special "unittest" path. Any such path is a special behavior that might break one of the class' assumption.

Take the example of the `version (unittest)` constructor: having a different constructor might break some assumptions the class makes (now or later), by -e.g.- leaving some fields empty.

Geod24 commented on source/agora/node/Network.d:
```diff
@@ -64,33 +65,53 @@ public class NetworkManager
     private const(string)[] dns_seeds;
 
+    /// no-op. for use with unittests only
+    version (unittest) public this () { }
```
Please no special-path for unittests :(

Geod24 reviewed (changes requested):
Remove the `unittest` special casing and we're G2G

Geod24 commented:
Sorry for the delay!

AndrejMitrovic commented:
Please just show me how you want it done?
